drivers/adc/ad7091r5/ad7091r5.c

```diff
@@ -0,0 +1,363 @@
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "ad7091r5.h"
+#include <string.h>
+#include "stdio.h"
+#include "stdlib.h"
+#include "error.h"
+#include "delay.h"
+#include "util.h"
+
+/******************************************************************************/
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+
+/**
+ * Read from device.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t ad7091r5_i2c_reg_read(struct ad7091r5_dev *dev,
+			      uint8_t reg_addr,
+			      uint16_t *reg_data)
+{
+	uint8_t buf[2];
+	int32_t ret;
+
+	ret = i2c_write(dev->i2c_desc, &reg_addr, 1, 1);
+	if (ret < 0)
+		return ret;
+
+	ret = i2c_read(dev->i2c_desc, buf, 2, 1);
+	if (ret < 0)
+		return ret;
+
+	*reg_data = (buf[0] << 8) | buf[1];
+
+	return ret;
+}
+
+/**
+ * Write to device.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t ad7091r5_i2c_reg_write(struct ad7091r5_dev *dev,
+			       uint8_t reg_addr,
+			       uint16_t reg_data)
+{
+	uint8_t buf[3];
+
+	buf[0] = reg_addr;
+	buf[1] = (reg_data & 0xFF00) >> 8;
+	buf[2] = reg_data & 0xFF;
+
+	return i2c_write(dev->i2c_desc, buf, ARRAY_SIZE(buf), 1);
+}
+
+/**
+ * Multibyte read from device. A register read begins with the address
+ * and autoincrements for each aditional byte in the transfer.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @param count - Number of bytes to read.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t ad7091r5_i2c_reg_read_multiple(struct ad7091r5_dev *dev,
+				       uint8_t reg_addr,
+				       uint8_t *reg_data,
+				       uint16_t count)
+{
+	uint8_t buf[512];
+	int32_t ret;
+
+	if (count > 512)
+		return FAILURE;
+
+	buf[0] = reg_addr;
+	memset(&buf[1], 0x00, count - 1);
+
+	ret = i2c_write(dev->i2c_desc, buf, 1, 0);
+	if (ret < 0)
+		return ret;
+
+	ret = i2c_read(dev->i2c_desc, buf, count, 0);
+	if (ret < 0)
+		return ret;
+
+	memcpy(reg_data, buf, count);
+
+	return ret;
+}
+
+
+/**
+ * I2C read from device using a mask.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - The mask.
+ * @param data - The register data.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t ad7091r5_i2c_read_mask(struct ad7091r5_dev *dev,
```

Comment:
I think this function is pointless, the caller can apply the mask easily.

Reply:
I would keep it for symmetry with  `ad7091r5_i2c_write_mask`.
Also there are a few projects where is a similar implementation for SPI:
https://github.com/analogdevicesinc/no-OS/blob/20c654149fcaa86349d6130ae5eebd56966020df/drivers/adc/ad7768-1/ad77681.c#L184